### ISSUE TITLE
Error enabling TDX 1.5 on Emerald Rapids/Fox Creek Pass

### ISSUE THREAD

**Issue by marcelamelara** [canonical/tdx#135]
Hello! I've run into the following error when trying to enable TDX 1.5 on an Emerald Rapids server (Fox Creek Pass). The server is running BIOS version 01.02.0001.

Here's the output of `sudo dmesg | grep -i tdx` after following steps 4.1-4.3 of your Host OS setup on the current tip of the noble-24.04 branch ([a8261b1](https://github.com/canonical/tdx/commit/a8261b15784a5bd9962bbc5af63d27444b8bbb3c)):

```
[    0.000000] Command line: BOOT_IMAGE=/boot/vmlinuz-6.8.0-1004-intel root=UUID=135651d7-8961-4488-ac65-186eef7498f1 ro kvm_intel.tdx=1 nohibernate
[    0.957299] Kernel command line: BOOT_IMAGE=/boot/vmlinuz-6.8.0-1004-intel root=UUID=135651d7-8961-4488-ac65-186eef7498f1 ro kvm_intel.tdx=1 nohibernate
[    1.630945] virt/tdx: BIOS enabled: private KeyID range [32, 64)
[    1.630948] virt/tdx: Disable ACPI S3. Turn off TDX in the BIOS to use ACPI S3.
[   10.444055] virt/tdx: TDX module: attributes 0x0, vendor_id 0x8086, major_version 1, minor_version 5, build_date 20231008, build_num 595
[   10.444059] virt/tdx: CMR: [0x100000, 0x6f800000)
[   10.444061] virt/tdx: CMR: [0x100000000, 0x107a000000)
[   10.444061] virt/tdx: CMR: [0x1080000000, 0x207c000000)
[   10.444062] virt/tdx: CMR: [0x2080000000, 0x307c000000)
[   10.444063] virt/tdx: CMR: [0x3080000000, 0x407c000000)
[   10.774047] virt/tdx: initialization failed: TDMR [0x0, 0x80000000): reserved areas exhausted.
[   10.824954] virt/tdx: module initialization failed (-28)
[   10.824960] kvm_intel: Failed to initialize TDX module.
```

Per https://github.com/canonical/tdx/issues/83#issuecomment-2151189710, it seems like the TDX module version in this BIOS is out of date with what the kernel expects/supports. I'll try to perform the suggested TDX module upgrade, though it'll take me at least a few days.

Thanks for any insights/recommendations in the meantime.

**Comment by kaihuang** [comment]
Thanks for reporting.  I will work on the fix and send to community for review, hopefully next week.

At the meantime, could you please provide the full dmesg log?

**Comment by hector-cao** [comment]
Thanks @kaihuang for this valuable feedback

**Comment by marcelamelara** [comment]
Thanks @kaihuang and @hector-cao ! I've attached my full dmesg log.
[06062024-dmesg.log](https://github.com/user-attachments/files/15644865/06062024-dmesg.log)

**Comment by kaihuang** [comment]
Hi, I updated some patches to https://github.com/intel/tdx/commits/tdxfix/

Which should fix this issue.

Could you please try to apply the latest 8 patches from commit

    73f253f x86/virt/tdx: Rename _offset to _member for TD_SYSINFO_MAP() macro

to the last commit:

    0d3288b x86/virt/tdx: Exclude memory region hole within CMR as TDMR's reserved area

to see whether this issue can be fixed?

If they cannot be applied to cleanly to your tree, please provide me your code base and I can regenerate the patchset.

**Comment by marcelamelara** [comment]
@kaihuang thank you so much. I will try applying these patches later today and get back to you on the results.

**Comment by kaihuang** [comment]
Hi,  It seems you are using some old TDX code so it's likely you won't be able to apply the patches cleanly.  I just updated the tdxfix branch to include a TEST patch just to test the TDX module initialization. 

You can still try to grab the 8 patches as mentioned above and try to apply on your local.  If that doesn't work, you can clone https://github.com/intel/tdx/commits/tdxfix/ and use that tree just to verify this bug can be fixed.  

How to test:

1) Build https://github.com/intel/tdx/commits/tdxfix/ with CONFIG_INTEL_TDX_HOST enabled and CONFIG_KVM_INTEL enabled.
2) just try to create a normal guest.  This will trigger TDX module initialization.  And you can see whether it is successful.

**Comment by marcelamelara** [comment]
@kaihuang Thanks for the instructions. I ended up trying to build directly form the `tdxfix` branch with the configurations you indicated, but unfortunately still ended up with a `make` error.

I'm building from commit ` 9dcbba00c21` with command ` make -j 64`. I'm not sure if I'm missing a configuration, I've attached my `.config` file.

**Comment by kaihuang** [comment]
@marcelamelara The commit you tried is correct.  

I just tried your config and it can build at my side.  I am using ubuntu 22.04 LTS.

Can you paste the error you met?

Btw, just in case, attached my .config, which is just based on the one from you (maybe slightly different though due to the build process).  
[config-kai.txt](https://github.com/user-attachments/files/15829600/config-kai.txt)

**Comment by marcelamelara** [comment]
@kaihuang I was able to successfully build, install and boot with your kernel patches on my Ubuntu 24.04 LTS machine. I also upgraded the TDX module firmware. 

While the TDX module does initialize now, I still get enablement errors when I go to launch a new TD guest. I'm using the default guest image included in commit `4e3c0e40a255`.

I've included my dmesg log and `tdvirsh new` output.

`$ ./tdvirsh new`:

```
Create and run new virsh domain from /home/mmelara/tdx/guest-tools/trust_domain.xml.template
---
error: Failed to define domain from /var/tmp/tdvirsh//tdvirsh.xml
error: XML error: Invalid value for attribute 'type' in element 'launchSecurity': 'tdx'.
error: failed to get domain 'tdvirsh'
error: failed to get domain 'tdvirsh'
error: failed to get domain 'tdvirsh-trust_domain-'
error: failed to get domain 'tdvirsh-trust_domain-'
```

`$ sudo dmesg | grep -i tdx`:

```
[    0.000000] Command line: BOOT_IMAGE=/boot/vmlinuz-6.9.0-rc5+ root=UUID=135651d7-8961-4488-ac65-186eef7498f1 ro kvm_intel.tdx=1 nohibernate
[    0.625544] Kernel command line: BOOT_IMAGE=/boot/vmlinuz-6.9.0-rc5+ root=UUID=135651d7-8961-4488-ac65-186eef7498f1 ro kvm_intel.tdx=1 nohibernate
[    1.301221] virt/tdx: BIOS enabled: private KeyID range [32, 64)
[    1.301225] virt/tdx: Disable ACPI S3. Turn off TDX in the BIOS to use ACPI S3.
[  855.210148] kvm: kvm_arch_post_init_vm: try to enable TDX
[  855.210248] virt/tdx: Production module.
[  855.210249] virt/tdx: TDX_FEATURES0: 0x1f0fbf
[  855.210250] virt/tdx: Module version: 1.5.05.00.0698, build_date: 20240129
[  855.210252] virt/tdx: Convertible Memory Regions (CMRs):
[  855.210253] virt/tdx:        CMR[0]: [0x100000, 0x6f800000)
[  855.210255] virt/tdx:        CMR[1]: [0x100000000, 0x107a000000)
[  855.210256] virt/tdx:        CMR[2]: [0x1080000000, 0x207c000000)
[  855.210257] virt/tdx:        CMR[3]: [0x2080000000, 0x307c000000)
[  855.210257] virt/tdx:        CMR[4]: [0x3080000000, 0x407c000000)
[  856.704482] virt/tdx: 1050644 KB allocated for PAMT
[  856.704489] virt/tdx: module initialized
[  856.704492] kvm: kvm_arch_post_init_vm: tdx_enable() returns 0
[  856.936161] kvm: kvm_arch_post_init_vm: try to enable TDX
[  856.936168] kvm: kvm_arch_post_init_vm: tdx_enable() returns 0
[  857.100517] kvm: kvm_arch_post_init_vm: try to enable TDX
[  857.100524] kvm: kvm_arch_post_init_vm: tdx_enable() returns 0
[  864.553376] kvm: kvm_arch_post_init_vm: try to enable TDX
[  864.553387] kvm: kvm_arch_post_init_vm: tdx_enable() returns 0
[  865.715434] kvm: kvm_arch_post_init_vm: try to enable TDX
[  865.715442] kvm: kvm_arch_post_init_vm: tdx_enable() returns 0
[  865.833630] kvm: kvm_arch_post_init_vm: try to enable TDX
[  865.833641] kvm: kvm_arch_post_init_vm: tdx_enable() returns 0
[  866.749972] kvm: kvm_arch_post_init_vm: try to enable TDX
[  866.749982] kvm: kvm_arch_post_init_vm: tdx_enable() returns 0
[  866.860042] kvm: kvm_arch_post_init_vm: try to enable TDX
[  866.860049] kvm: kvm_arch_post_init_vm: tdx_enable() returns 0
[  868.377187] kvm: kvm_arch_post_init_vm: try to enable TDX
[  868.377196] kvm: kvm_arch_post_init_vm: tdx_enable() returns 0
[ 1019.207315] kvm: kvm_arch_post_init_vm: try to enable TDX
[ 1019.207322] kvm: kvm_arch_post_init_vm: tdx_enable() returns 0
```

Thanks for any input on this!

**Comment by kaihuang** [comment]
Hi @marcelamelara the purpose of https://github.com/intel/tdx/commits/tdxfix/ is to verify my patches can actually fix this issue.  If you were using it to create TDX guest, then it is expected to fail.

To fix this issue in your TDX host kernel, I need to access your tree so that I can generate a patch for you.

Could you let me know how to access your TDX host kernel code?

**Comment by hyuan3** [comment]
HI @kaihuang , I will work with Canonical team to port your patches and do internal validation. Then Canonical team can merge it to Ubuntu kernel.

**Comment by fanchen2** [comment]
As we cannot reproduce this issue on our test machine, @hector-cao could you provide a tmp build for the reporter to verify this issue after @hyuan3 finish the patch backport?

**Comment by marcelamelara** [comment]
> Could you let me know how to access your TDX host kernel code?

I actually installed the kernel built through the `tdxfix` patches (I ended up with kernel 6.9.0-rc5+), and that's what I'm running on the host now. Is there a different host kernel I should have been using?

**Comment by hector-cao** [comment]
@marcelamelara I m working on providing you a debug kernel out of the box for you to test our TDX solution, is that ok ?

**Comment by hector-cao** [comment]
@marcelamelara I have a kernel with the fix applied, could you please give a try ? Please checkout this branch https://github.com/canonical/tdx/tree/dev-enable-testing-ppa and run

```
sudo ./scripts/enable-kernel-testing.sh
```

after reboot, you should see this kernel version

```
$ uname -r
6.8.0-1006-intel
```

**Comment by marcelamelara** [comment]
@kaihuang Got it! Thanks for clarifying. 

@hector-cao Thanks so much for the debug kernel. I'll give it a try today.

**Comment by marcelamelara** [comment]
@hector-cao I've successfully installed and booted into the `6.8.0-1006-intel`, thank you. Should I go ahead and try to launch a TD guest now? From which branch/commit? Thanks!

**Comment by hector-cao** [comment]
You can use the same branch (dev-enable-testing-ppa), please follow https://github.com/canonical/tdx?tab=readme-ov-file#51-create-a-new-td-image to create an image and https://github.com/canonical/tdx?tab=readme-ov-file#6-boot-td to run a TD . In the meantime, can you grub the host dmesg (```sudo dmesg | grep tdx```) output and paste it here please

**Comment by marcelamelara** [comment]
Here's the latest host dmesg output:

```
[    0.000000] Command line: BOOT_IMAGE=/boot/vmlinuz-6.8.0-1006-intel root=UUID=135651d7-8961-4488-ac65-186eef7498f1 ro kvm_intel.tdx=1 nohibernate
[    0.624877] Kernel command line: BOOT_IMAGE=/boot/vmlinuz-6.8.0-1006-intel root=UUID=135651d7-8961-4488-ac65-186eef7498f1 ro kvm_intel.tdx=1 nohibernate
[    1.289113] virt/tdx: BIOS enabled: private KeyID range [32, 64)
[    1.289117] virt/tdx: Disable ACPI S3. Turn off TDX in the BIOS to use ACPI S3.
[    8.222925] virt/tdx: TDX module: attributes 0x0, vendor_id 0x8086, major_version 1, minor_version 5, build_date 20240129, build_num 698
[    8.222929] virt/tdx: CMR: [0x100000, 0x6f800000)
[    8.222930] virt/tdx: CMR: [0x100000000, 0x107a000000)
[    8.222933] virt/tdx: CMR: [0x1080000000, 0x207c000000)
[    8.222934] virt/tdx: CMR: [0x2080000000, 0x307c000000)
[    8.222935] virt/tdx: CMR: [0x3080000000, 0x407c000000)
[    9.355881] virt/tdx: 1050644 KB allocated for PAMT
[    9.355890] virt/tdx: module initialized
```

I'm going to try to run a TD guest next.

**Comment by marcelamelara** [comment]
@hector-cao I followed steps 5.1 and 6.1 to create and boot a TD exactly, but got the following error:

```
qemu-system-x86_64: -object tdx-guest,id=tdx: Parameter 'qom-type' does not accept value 'tdx-guest'
```

**Comment by hector-cao** [comment]
```dpkg -l | grep qemu-system-x86``` Please

**Comment by marcelamelara** [comment]
The output of `dpkg -l | grep qemu-system-x86`:

```
ii  qemu-system-x86                       1:8.2.2+ds-0ubuntu1                              amd64        QEMU full system emulation binaries (x86)
```

Would you like me to share the /tmp/tdx-guest-setup.txt log as well?

**Comment by hector-cao** [comment]
The qemu you are using is not the right one, did you run ```setup-tdx-host.sh``` before ```enable-kernel-testing.sh``` ?

**Comment by hector-cao** [comment]
You can try to run ```setup-tdx-host.sh``` followed by ```enable-kernel-testing.sh``` again

**Comment by marcelamelara** [comment]
I ran only `setup-tdx-host.sh` and that seems to have fixed the QEMU issue. I was able to boot into the TD now and verify the TD (step 7). Thanks so much!

**Comment by hector-cao** [comment]
Can you double check you still have the right kernel ? ```uname -r```

**Comment by marcelamelara** [comment]
Yes, confirmed, I'm still running the `6.8.0-1006-intel` kernel. I haven't rebooted, though.

**Comment by hector-cao** [comment]
Ok, if you reboot, you will get back to the previous kernel, do you need further assistance or can we close this issue ?

**Comment by marcelamelara** [comment]
No, things are running as expected now. I have a workaround to make sure grub boots into the patched kernel. Thanks again so much for all your support on this!

**Comment by hector-cao** [comment]
@marcelamelara For the moment, we decided to not bring the fix in out current release on 24.04, however we will keep it in mind for the next release (24.10). Thanks !

**Comment by marcelamelara** [comment]
> For the moment, we decided to not bring the fix in out current release on 24.04, however we will keep it in mind for the next release (24.10)

Thanks for the heads up. Does this mean the fix won't be under LTS?

**Comment by ThibF** [comment]
linux-intel kernel used for this tdx project isn't going to be maintained the traditional support duration and we are going to follow a rolling model.
About this specific issue, in fact we are going to release a fix, you can expect soon (~2weeks) to be released, testing is ongoing on it, you can find it here : https://launchpad.net/~kobuk-team/+archive/ubuntu/tdx-proposed

**Comment by kaihuang** [comment]
Just let you guys know that I've send an updated version to community:

https://lore.kernel.org/kvm/cover.1721186590.git.kai.huang@intel.com/T/

And since this issue only appears on some large machine, in case it's not easy to find a testing machine, we can use the 'memmap' kernel command line to test the fix on any TDX enabled machine.  

E.g., the 'memmap' command line I am using is:

memmap='1M#2M,1M#4M,1M#6M,1M#8M,1M#10M,1M#12M,1M#14M,1M#16M,1M#18M,1M#20M,1M#22M,1M#24M,1M#26M,1M#28M,1M#30M,1M#32M'

**Comment by bktan8** [comment]
LP#: https://bugs.launchpad.net/kobuk/+bug/2072717

**Comment by bktan8** [comment]
Resolved in release 2.1.  https://github.com/canonical/tdx/releases/tag/2.1.
